[AG-15555] Feature(ssrm): add support for expand all collapse all

sdwvit opened the pull request:
- Add a prop to gridOptions to enable expandAll/collapseAll on all rows, not just loaded ones
 - Update error messages to include feature context for `getRowId` validation
 - Document changed behavior in ag-grid-docs for server-side grouping
 - bonus: Allows using optional props in warning/error messages

Fix [AG-15555](https://ag-grid.atlassian.net/browse/AG-15555)
Fix [AG-15693](https://ag-grid.atlassian.net/browse/AG-15693)

<img width="958" height="1149" alt="Screenshot 2025-08-12 at 14 28 34" src="https://github.com/user-attachments/assets/3751399d-c553-4272-85c2-5612b44eafc2" />


<img width="1301" height="589" alt="Screenshot 2025-08-12 at 11 13 28" src="https://github.com/user-attachments/assets/12f40bd6-6f78-47b1-8290-a75ae89e4459" />

Commit:
Add state type

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Refactor expandAll to use transactional callback for row expansion

- Update method signature to accept callback instead of boolean value
- Replace direct expansion logic with forEachNode traversal for atomic updates
- Handle stub nodes explicitly during expansion traversal
- Track expanded/collapsed state in service's toggledNodes map
- Clear queued rows on column group changes to prevent stale state
- Migrate expandAll implementation from row model to expansion service

Signed-off-by: Victor Musienko <victor.m@noibu.com>

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,21 +11,27 @@ import type {
         this.addManagedEventListeners({
             columnRowGroupChanged: () => {
-                this.queuedRowIds.clear();
+                this.expandedState.toggledNodes.clear();
```
Changing this would be breaking but food for thought: (don't make any change, but maybe worth looping in Zoheil and asking what he thinks/if we should do something about this)
What if I went from grouping Country, to grouping Country and Sport, is it right to clear all row grouping state or should we try to maintain because we've simply added a new deeper dimension which doesn't impact which groups are available at the top level?

sdwvit replied:
in your opinion what should happen here? I am not sure I understand this event fully to make a call

AndrewGlazier replied:
In my opinion this is wrong:
1/ Open https://plnkr.co/edit/cdg73coW29MrN3Bh?open=main.js
2/ Right click group by Sport
Expected: Only the contents of the country groups need refreshed
Actual: Both the expanded state of the country groups is lost, and the entire grid is refreshed

There might be some niche edge cases where aggregations would be impacted and so need the entire refresh, so it'd probably need to be behind a flag, and it'd be breaking to change this by default anyway.

We don't need to make a judgement call here, just raise the issue/scenario with Zoheil and let him do so

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -59,18 +65,15 @@ export class ServerSideExpansionService extends BaseExpansionService implements
 
     public expandRows(rowIdsToExpand: string[], rowIdsToCollapse?: string[]): void {
-        const { serverSideRowModel, queuedRowIds } = this;
         const processNodes = (rowIds: string[], expanded: boolean) => {
```
This is coming from initialState/stateService, we probably need to also store selectAll in our initial state now (and probably find a backwards-compatible way to handle old states)

Note when you fix AG-15693 we also need to somehow store _that_ state (i.e what nodes are expanded by default vs which ones have been explicitly closed, even if expandAll is false)

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -35,8 +41,8 @@ export class ServerSideExpansionService extends BaseExpansionService implements
-        if (this.queuedRowIds.has(rowNode.id!)) {
-            this.queuedRowIds.delete(rowNode.id!);
+        const shouldExpand = this.expandedState.expandAll !== this.expandedState.toggledNodes.has(rowNode.id!);
+        if (shouldExpand) {
```
I've opened AG-15693 to tackle an issue related to this - don't fix now as we can keep this PR small, but I've assigned you that ticket

sdwvit replied:
I think this would be useful to fix that issue here too, as the approach to the whole methodology is different

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,21 +11,27 @@ import type {
 export class ServerSideExpansionService extends BaseExpansionService implements NamedBean, IExpansionService {
     beanName = 'expansionSvc' as const;
 
+    private expandedState: ExpansionState = { expandAll: false, toggledNodes: new Set() };
```
I can't see a request for this, but if possible it'd be a good idea to add set/get state and serialize this - as it'll be a good way for users to update the expanded rows - if you have time can we scope creep this?

I'm not sure why it's not a ticket already

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -82,8 +85,17 @@ export class ServerSideExpansionService extends BaseExpansionService implements
+                this.expandedState.toggledNodes.add(node.id!);
+            } else {
+                if (node.hasChildren()) {
+                    node.setExpanded(value);
```
This pattern isn't great - I know it's not yours and this is scope creep, but if you want to implement an important fix:
Look at how setExpanded works, (calls setRowNodeExpanded on the parent class) it dispatches a `rowGroupOpened` event synchronously every time
This event is listened to by lots of things, stateService (which hacks in a debounce), editService (which doesn't debounce it at all) and the `expandListener` (SSRM only)
Every time we do this, we subsequently fire a `storeUpdated` event which then recomputes all of the row indexes in the SSRM synchronously (which is very inefficient)

If you fancy taking up the task (we can do in a different PR on the same ticket if you want) it'd be good to introduce a new pattern for bulk expanding rows and detach the internal grid logic from the synchronous, per-row `rowGroupOpened` event


For expand-all this currently works fine due to the flag we have, but that's not a great pattern and we don't offer users any way to bulk expand rows (that isn't expandAll) - so when doing the setter/getter we probably want to improve this

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts:
```diff
@@ -464,20 +465,12 @@ export class ServerSideRowModel extends BeanStub implements NamedBean, IServerSi
     }
 
-    public expandAll(value: boolean): void {
+    public expandAllTransactional(cb: Parameters<IRowModel['forEachNode']>[0]): void {
```
Don't love the change in this type, but have a look at the other comments and see if you can change this back as a result of any of them, if you don't feel you can then it's fine

sdwvit replied:
yea I also feel that, will replace it with something nicer and less coupled

AndrewGlazier reviewed (commented):
A few notes

sdwvit committed:
wip

sdwvit committed:
feat(server-side-row-model): update method names and state structures

- Rename expandAllTransactional to forEachNodeTransactional for clarity
- Replace Set with Record<> for toggledNodes and interactedWith state management
- Add resetExpandedState method to reinitialize expansion state
- Update expandAll implementation to use forEachNodeTransactional
- Simplify expandedState tracking with plain object instead of Proxy

sdwvit committed:
feat(expansion): refactor state management with Proxies and simplify expansion logic

- Removed conditional expand check in blockUtils.ts, directly set via expansion service
- Introduced Proxies in ServerSideExpansionService to track toggledNodes/interactedWith interactions
- Replaced DEFAULT_EXPANDED_STATE with factory function using Proxies for memory efficiency
- Removed resetExpandedState method, use getDefaultExpansionState in expandAll
- Updated isExpanded logic to use Proxy-based state checks

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
+    /** Base expansion state for all nodes, whether they have been loaded or not */
+    expandAll: boolean;
+    /** Whether the user has interacted with all nodes */
+    interactedWithAll: boolean;
```
Is this necessary, can we just merge this into `expandAll`

I.e, expandAll: boolean | null, where if expandAll == null it's not been interacted with, however if it's true/false we know it's be explicitly toggled

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
         this.beans.eventSvc.dispatchEvent({
             type: 'expandOrCollapseAll',
-            source: value ? 'expandAll' : 'collapseAll',
+            source: expanded ? 'expandAll' : ' collapseAll',
```
' collapseAll' looks like a typo

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
 import { BaseExpansionService } from '../../rowHierarchy/baseExpansionService';
 import type { ServerSideRowModel } from '../serverSideRowModel';
 
+interface ExpansionState {
```
This could probably be a stricter type, which may make it easier to follow the behaviour for anyone unfamiliar with this, e.g:
```
interface AllExpandedState {
   expandAll:  boolean;
   toggledNodes: string[]; // I'm aware you're currently using a map for quick lookup, could be a set internally and arr exported though
   // touchedNodes not required here as it's implicit because expandAll is non-null
}

interface GranularExpandedState {
  expandAll: null; // expand all boolean/null now becomes something we can target on ExpansionState for drilling into the type
  toggledNodes: string[];
  touchedNodes: string[];
}

type ExpansionState = AllExpandedState | GranularExpandedState;
```

(This is an example of the exported state, I imagine internally we'd need to use Sets where arrays are used)

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
+    /** Whether the user has interacted with all nodes */
+    interactedWithAll: boolean;
+    /** RowNode IDs of those nodes whose expansion state differs from the base state */
+    toggledNodes: Record<string, boolean | undefined>; // in reality this is true | undefined
```
For internal lookup, any reason not to use a set? I'm assuming real type is Record<string, true> rather than Record<string, true | undefined> so this could be a Set<string> (same with interactedWith below)

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
+        interactedWith: new Proxy({} as ExpansionState['interactedWith'], {
+            get: (target, p) => {
+                const rowIdC = p as keyof typeof target;
+                return self.interactedWithAll ? delete target[rowIdC] || true : target[rowIdC];
```
This should never happen right? Because you create a fresh instance of the expansion state when expandAll is toggled, and your setter won't allow setting new values if interactedWithAll is set

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
+ * Using a proxy here allows us to automagically update the `interactedWith` state,
+ * which is used to determine the default expansion state of the nodes. And be memory conscious as well.
+ */
+const getDefaultExpansionState = (defaultOverrides: Partial<ExpansionState> = {}) => {
```
I'm not a fan of the proxy here because I think it adds some complexity that isn't really needed, it will make it harder for people to follow/debug the state of this object as it changes, whereas in this instance I think a simple setter func could handle this logic and be a bit cleaner, thoughts?

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/index.mdoc:
```diff
@@ -134,7 +134,9 @@ api.collapseAll();
 
-Calling `expandAll()` and `collapseAll()` will have no impact on rows yet to be loaded.
+{% note %}
+Calling `expandAll()` and `collapseAll()` will apply to all rows, including those not yet loaded.
```
We won't actually be able to do this by default I think, we might need to introduce a new grid option flag due to this being a breaking change, even if the old behaviour is a bit rubbish

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,127 @@ import type {
     }
 
-    public checkOpenByDefault(rowNode: RowNode): void {
+    public checkOpenByDefault(rowNode: RowNode): boolean {
```
Might be worth a rename on this one; isRowExpanded may be more applicable now the signature has changed

sdwvit committed:
refactor expansion state management to use Sets and improve toggle logic

- Replaced Proxy-based state tracking with Set structures for toggledNodes and interactedWith
- Introduced toggleNode method to handle expansion state changes
- Simplified isExpanded logic using Set operations
- Renamed checkOpenByDefault to isRowExpanded for clarity
- Added expandAll method to manage global expansion state

sdwvit committed:
simplify state

sdwvit committed:
feat(serverSideRowModel): add server-side row expansion state management

- Added IServerSideRowExpansionState interface for tracking expand/collapse state
- Introduced getServerSideExpandState and setServerSideExpandState APIs
- Updated ServerSideExpansionService to persist expansion state across sessions
- Added GroupFlags.usePre34Behaviour to control expansion behavior consistency
- Modified expansion service to handle toggled nodes and expandAll state separately
- Added type exports for expansion state interfaces in main module

sdwvit committed:
add a todo comment

sdwvit committed:
- allow optional _warn arguments with default value

sdwvit committed:
- generalize error 188

sdwvit committed:
feat(grid): update group behaviour flag and conditional logic

- Rename `usePre34Behaviour` to `useNewBehaviour` in grid options
- Add `groupFlags` to internal grid option array for tracking
- Update warning message to differentiate between selection and grouping features
- Replace old flag reference with new behaviour flag in expansion service logic
- Adjust expansion conditionals to use new behaviour flag for group state management

sdwvit committed:
Update expansion state interface and implementation to use enum for expandAll

sdwvit committed:
revert state work

sdwvit committed:
revert state work

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555' into sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555

sdwvit committed:
Update variable name and logic for row expansion state management

- Renamed variable from isExpandAll to expandAllStatus for clarity
- Updated isExpanded logic to use expandAllStatus instead of isExpandAll
- Modified hasInteractedWithAll to check expandAllStatus !== 'notInteracted'
- Updated reset method to assign expandAllStatus instead of isExpandAll

sdwvit committed:
lint

sdwvit committed:
Rename groupFlags.useNewBehaviour to useSsrmNewBehaviour

- Update documentation to reference new flag name in expandAll/collapseAll notes
- Rename groupFlags.useNewBehaviour to useSsrmNewBehaviour in gridOptions.ts
- Update serverSideExpansionService.ts to use new flag variable name
- Adjust enterprise row model check to useSsrmNewBehaviour instead of useNewBehaviour

sdwvit committed:
doc

sdwvit committed:
feat(grid): export GroupFlags type for external usage

Commit:
bundle size bump

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Merge remote-tracking branch 'origin/sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555' into sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555

AndrewGlazier commented on packages/ag-grid-community/src/entities/gridOptions.ts:
```diff
@@ -2121,6 +2121,11 @@ export interface GridOptions<TData = any> {
+     * Controls how groups behave.
+     * @agModule `RowGroupingModule` / `TreeDataModule`
+     */
+    groupFlags?: GroupFlags;
```
We tend to use flat grid options (mostly) as it makes for easier config for users. E.g for this, if the user did:
<AgGridReact groupFlags={{ useSsrmNewBehaviour: true }} />

Then every time their component re-rendered the grid would need to perform reactivity updates, so they'd have to extract and memoise this object to reduce unwanted side effects and processing.

I'd suggest a name more like:
ssrmExpandAllState (though maybe check with Zoheil to confirm that we really do need a flag if you haven't already) (Open to better names still as this isn't an easy thing to name)

sdwvit replied:
ssrmExpandAllState doesn't indicate boolean nature of this flag. it can only be true/false. undefined is coerced to false. I'll think a bit more about the name here.

sdwvit replied:
how about `ssrmExpandAllAffectsAllRows`?

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/index.mdoc:
```diff
@@ -134,7 +134,9 @@ api.collapseAll();
 Calling `expandAll()` and `collapseAll()` will impact **all loaded group nodes**, including those not visible due to their containing group been closed. This means there could potentially be a huge number of groups expanded, so this method should be used very wisely to not create massive amount of server requests and loading a large amount of data.
 
-Calling `expandAll()` and `collapseAll()` will have no impact on rows yet to be loaded.
+{% note %}
```
We'll want to introduce a new example which demonstrates the new feature, and the difference of enabling and suppressing this flag

e.g a checkbox for toggling the value. (similar to this https://www.ag-grid.com/javascript-data-grid/grouping-data/#hiding-parents-of-individual-rows)

An expand and collapse all button above the grid

and appropriate properties (cacheBlockSize = 0, purgeClosedRowNodes = true) to make the difference apparent when doing expand/collapse all

sdwvit replied:
added, thnx!

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/serverSideRowModel.ts:
```diff
@@ -157,7 +157,10 @@ export class ServerSideRowModel extends BeanStub implements NamedBean, IServerSi
-            _warn(188);
+            _warn(188, { feature: 'selection' });
+        }
+        if (this.gos.get('groupFlags')?.useSsrmNewBehaviour && !this.gos.exists('getRowId')) {
```
This check would probably be better placed in `gridOptionsValidations` as people can drop the validations module for their production build (minor bundle size improvement)

AndrewGlazier commented on packages/ag-grid-enterprise/src/serverSideRowModel/services/serverSideExpansionService.ts:
```diff
@@ -11,82 +11,99 @@ import type {
 import type { ServerSideRowModel } from '../serverSideRowModel';
 
+interface IServerSideRowExpansionState {
+    expandAll: 'notInteracted' | 'expandAll' | 'collapseAll';
```
minor nit: don't feel the need to change this if you disagree, but as we're going to be expecting users to serialize this doing a union of undefined/boolean would be less storage impact on our users

Minor, but food for thought

Commit:
Rename groupFlags to ssrmExpandAllAffectsAllRows and update related logic

- Added new `ssrmExpandAllAffectsAllRows` grid option for controlling expand/collapse behavior
- Updated documentation to reflect renamed configuration flag
- Modified server-side expansion logic to use new flag instead of deprecated property
- Removed `groupFlags` reference from property keys list
- Enhanced validation to enforce `getRowId` when using `ssrmExpandAllAffectsAllRows`
- Cleaned up warning message related to group flags in row model initialization

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Merge branch 'latest' into sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555

# Conflicts:
#	testing/module-size/moduleDefinitions.ts

Commit:
- Updated documentation to explain how `expandAll()`/`collapseAll()` interacts with this flag
- Introduced example implementation demonstrating the new functionality in server-side grouping scenarios

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
- update example

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
- improve getRowId

Signed-off-by: Victor Musienko <victor.m@noibu.com>

AndrewGlazier commented on packages/ag-grid-community/src/entities/gridOptions.ts:
```diff
@@ -2127,6 +2127,13 @@ export interface GridOptions<TData = any> {
+     * and interacting with the group overrides the default expansion state set by `isServerSideGroupOpenByDefault`.
+     * @agModule RowGroupingModule / TreeDataModule
+     */
+    ssrmExpandAllAffectsAllRows?: boolean | undefined;
```
This needs to be added to documentation/ag-grid-docs/src/content/api-documentation/grid-options/properties.json so that it will show up on the api documentation page, this should also link back to the relevant docs section

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/_examples/expand-all-affects-all-rows/index.html:
```diff
@@ -0,0 +1,13 @@
+    <div class="example-header">
+        <label>
+            <span>Change gridOptions['ssrmExpandAllAffectsAllRows'] value:</span>
+            <select id="input-display-type" onchange="onOptionChange()">
```
As it's just true/false can you make this a checkbox

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/_examples/expand-all-affects-all-rows/index.html:
```diff
@@ -0,0 +1,13 @@
+<div class="example-wrapper" style="height: 100%">
+    <div class="example-header">
+        <label>
+            <span>Change gridOptions['ssrmExpandAllAffectsAllRows'] value:</span>
```
For consistency with other docs can you just make this `ssrmExpandAllAffectsAllRows` without the pseudo typescript, not everyone will be using TS and it's not a convention we've used elsewhere

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/index.mdoc:
```diff
@@ -134,7 +134,11 @@ api.collapseAll();
+To change the default behaviour of `expandAll()` and `collapseAll()` to expand or collapse **all group nodes**, not only loaded group nodes, set `ssrmExpandAllAffectsAllRows` flag in `GridOptions` to `true`.
+{% /note %}
+
+{% gridExampleRunner title="expand-all-affects-all-rows" name="expand-all-affects-all-rows"  exampleHeight=540 /%}
```
I don't think that the default docs for this Expand All / Collapse All section are very good (it's the same for all the SSRM docs, but focusing on this section), do you mind trying to refactor them? For context the Row Grouping (non SSRM docs) are an example of something I've entirely rewritten more recently, the SSRM docs are a bit outdated

We tend to go for something like this: https://www.ag-grid.com/react-data-grid/grouping-single-group-column/#checkbox-selection

Note how the style is:
Simple intro to the feature
Example
Small config example

We've made this change to remove some of the more rambling-style flow of documentation, and reduce the docs to key info.

I'd suggest something like:
```
Expand and Collapse All
Group rows can be expanded or collapsed using the api methods `expandAll` and `collapseAll`.  By default, this will only impact loaded group rows but enabling the `ssrmExpandAllAffectsAllRows` grid option changes this behaviour to instead expand groups as they are loaded.

{ EXAMPLE }

The example above is configured using the following code snippet:
const gridOptions = {
   ssrmExpandAllAffectsAllRows: true,
}
```

sdwvit replied:
good point let me do it

AndrewGlazier commented:
There's one additional thing we should take into account - when using column menus with the _client side_ row model, we show the Expand and Collapse all buttons in the menu.

For the SSRM we don't - I think when using the new property we should show expand/collapse all buttons the same as the CSRM.

sdwvit committed:
Add ssrmExpandAllAffectsAllRows property to grid options

- Added ssrmExpandAllAffectsAllRows property with description on expand/collapse behavior
- Requires getRowId function when enabled, linked to row-ids documentation
- Default set to false, overriding group expansion state

sdwvit committed:
Merge branch 'latest' into sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555

# Conflicts:
#	testing/module-size/moduleDefinitions.ts

sdwvit committed:
Update documentation and example for ssrmExpandAllAffectsAllRows configuration

 - Added explanation for ssrmExpandAllAffectsAllRows flag in expand/collapse docs
 - Updated example to use checkbox instead of select for config toggle
 - Adjusted grid height calculations in example UI
 - Fixed note about expandAll() behavior when flag is enabled
 - Updated TS config to dynamically read checkbox state
 - Improved documentation for server-side row model expansion controls

sdwvit committed:
Update ssrmExpandAllAffectsAllRows description in API docs

 - Added "(Server-side Row Model only)" prefix to clarify scope

AndrewGlazier commented on documentation/ag-grid-docs/src/content/docs/server-side-model-grouping/index.mdoc:
```diff
@@ -206,7 +211,7 @@ Then the columns are set up so that country uses a `valueGetter` that uses the f
 ## Filters
 
-By default the grid will fully purge the grid when impacted by the change in filters. The grid can be configured to only refresh when the group has been directly impacted by enabling `serverSideOnlyRefreshFilteredGroups`. Be aware, this can mean your grid may have empty group rows. This is because the grid will not refresh the groups above the groups it deems impacted by the filter.
+By default, changing filters fully purges the grid. Thought, it can be configured to only refresh when the group has been directly impacted by enabling `serverSideOnlyRefreshFilteredGroups`. Be aware, this can mean your grid may have empty group rows. This is because the grid does not refresh the groups above the groups it deems impacted by the filter.
```
Thought -> Though

AndrewGlazier commented on documentation/ag-grid-docs/src/content/api-documentation/grid-options/properties.json:
```diff
@@ -1427,6 +1427,13 @@
+            "description": "(Server-side Row Model only) Controls whether expandAll / collapseAll operations apply to all rows (not just loaded ones), and whether group interactions override the default expansion state set by `isServerSideGroupOpenByDefault`. When true, users must provide a `getRowId` function to ensure proper row identification. Default: `false`.",
+            "more": {
+                "name": "Row IDs",
+                "url": "./row-ids/"
```
Wrong url? This should link to the expand/collapse docs section

sdwvit replied:
row ids are required for this flag to work, but I can link both perhaps

sdwvit committed:
Fix typo and clarify server-side grouping filter behavior

 - Fix typo in "Thought" → "Though" in filtering documentation
 - Update link in ssrmExpandAllAffectsAllRows description to point to expand/collapse section

sdwvit committed:
docs(grid): update example comments for expand/collapse buttons

- Clarified button actions in example comments (first click expands, second collapses)
- Updated code example to show expand includes unloaded rows when flag is enabled
- Fixed button text labels to match functionality (Expand rows / Collapse rows)

sdwvit committed:
- Removed client-side row model check for expand/collapse.

sdwvit committed:
improve condition

sdwvit committed:
bump bundle size

sdwvit committed:
Merge branch 'latest' into sdwvit/feature/ssrm/add-support-for-expand-all-collapse-all/AG-15555

# Conflicts:
#	testing/module-size/moduleDefinitions.ts

sdwvit committed:
bundle size
